pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoyfilter
+
+import (
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/gogo/protobuf/proto"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+// ApplyClusterPatches applies patches to CDS clusters
+func ApplyClusterPatches(patchContext networking.EnvoyFilter_PatchContext, proxy *model.Proxy,
+	push *model.PushContext, clusters []*xdsapi.Cluster) []*xdsapi.Cluster {
+
+	envoyFilterWrappers := push.EnvoyFilters(proxy)
+	clustersRemoved := false
+	for _, efw := range envoyFilterWrappers {
+		for _, cp := range efw.Patches[networking.EnvoyFilter_CLUSTER] {
+			if cp.Operation != networking.EnvoyFilter_Patch_REMOVE &&
+				cp.Operation != networking.EnvoyFilter_Patch_MERGE {
+				continue
+			}
+			for i := range clusters {
+				if clusters[i] == nil {
+					// deleted by the remove operation
+					continue
+				}
+
+				if patchContextMatch(patchContext, cp) && clusterMatch(clusters[i], cp) {
+					if cp.Operation == networking.EnvoyFilter_Patch_REMOVE {
+						clusters[i] = nil
+						clustersRemoved = true
+					} else {
+						proto.Merge(clusters[i], cp.Value)
+					}
+				}
+			}
+		}
+
+		// Add cluster if the operation is add, and patch context matches
+		for _, cp := range efw.Patches[networking.EnvoyFilter_CLUSTER] {
+			if cp.Operation == networking.EnvoyFilter_Patch_ADD {
+				if patchContextMatch(patchContext, cp) {
+					clusters = append(clusters, cp.Value.(*xdsapi.Cluster))
```

Comment:
should we add some error handling around the case when `cp.Value` doesn't successfully merge or if the type assertion around `xdsapi.Cluster` panics?

Reply:
Thats handled in model/envoyfilter.go where we pre-convert things to their types.

Reply:
and in validation.go as well